### PR TITLE
Prevent post-completion state and resource leakage in goyek.A

### PR DESCRIPTION
**Pull request by pellared**
## Summary

Lifecycle-sensitive `A` helpers are now coordinated with task cleanup completion.

`A.Cleanup`, `A.WithContext`, `A.Setenv`, `A.TempDir`, and `A.Chdir` atomically check whether cleanup registration is still open. Resource-affecting calls rejected after cleanup has closed panic before changing process state. Calls admitted while completion is racing register any required cleanup before releasing the lifecycle lock.

The runner also waits for admitted calls to finish before producing its result. This ensures that:

- accepted cleanup callbacks are drained before the runner returns;
- environment, temporary-directory, and working-directory changes are restored;
- failures from admitted calls are reflected in the task result; and
- a rejected `WithContext` call cancels its newly derived context.

Regression tests cover calls after completion on original and derived `A` values, cleanup and environment operations racing with completion, admitted failures during cleanup, and relevant error and cleanup-failure paths. The changelog is updated.

This is a fail-fast safeguard for accidental violation of `A`'s existing lifecycle contract. Callers must still wait for all goroutines using `A` before the runner returns, and process-wide operations remain unsupported in parallel tasks.

---
*PR created automatically by Jules for task [1506413421333354607](https://jules.google.com/task/1506413421333354607) started by @pellared*

### CODE REVIEW

File: runner.go [hunk 1]
```diff
@@ -109,6 +110,10 @@ func (r taskRunner) run(in Input) Result {
 
 	finished, panicVal, panicStack := a.run(r.action)
 
+	a.mu.Lock()
+	*a.finished = true
```
Comment (pellared) [inline runner.go]: This transition is not atomic with `runCleanups` observing an empty cleanup slice. A late goroutine can append a `Cleanup` after the final empty check but before this lock is acquired; the append succeeds, this then marks the task finished, and no consumer remains to run that cleanup. Please set the terminal state under `a.mu` at the point the cleanup loop determines the queue is empty (including cleanup-panic paths), so registration is either accepted and drained or rejected.

File: a.go [hunk 2]
```diff
@@ -253,6 +258,13 @@ func (a *A) Cleanup(fn func()) {
 // Because Setenv affects the whole process, it should not be used in parallel tasks.
 func (a *A) Setenv(key, value string) {
 	a.Helper()
+	a.mu.Lock()
+	if a.finished != nil && *a.finished {
```
Comment (pellared) [inline a.go]: The check is separated from both the side effect and cleanup registration. If finalization wins after the unlock, `os.Setenv` can mutate the process and the later `Cleanup` can panic or be stranded, leaving the value behind. `TempDir` has the same create/leak window, and `Chdir` can run its restore cleanup before the later `os.Chdir`. Please serialize each check + side effect + cleanup append with the terminal transition (or use an equivalent in-flight-operation protocol). Be careful to unlock before `Fatal` paths and avoid re-entering `Cleanup`/`Setenv` while holding this mutex.

File: a_test.go [hunk 3]
```diff
@@ -211,6 +211,73 @@ func TestA_WithContext_concurrent_fail_derived(t *testing.T) {
 	assertEqual(t, got.Status, goyek.StatusFailed, "should return proper status")
 }
 
+func TestA_PostCompletionPanic(t *testing.T) {
+	var capturedA *goyek.A
+	res := goyek.NewRunner(func(a *goyek.A) {
+		capturedA = a
+	})(goyek.Input{})
```
Comment (pellared) [inline a_test.go]: This only tests calls made after the runner has already returned, so it cannot catch either completion-boundary race. Please add concurrency coverage for the invariant that every attempted cleanup/resource operation is either rejected before mutation or accepted and fully cleaned up. A successful `Cleanup` registration must never be silently left unexecuted.

**Review by pellared — COMMENTED** [review]
The current implementation does not yet support the PR body's concurrency and leak-prevention claims. The lifecycle transition and resource operations need to be linearized, and the tests need to cover the completion boundary. Also, the claim that these exact late-call panics match `testing.TB` is inaccurate: current `testing.Cleanup`, `Setenv`, `TempDir`, and `Chdir` do not perform this completion check (although the broader defensive intent is similar).

File: a.go [hunk 4]
```diff
@@ -238,109 +256,196 @@ func (a *A) Helper() {
...
+				restoreDir()
+			},
+		}
+	})
+
+	if parallel {
 		a.Fatalf("Chdir called in a parallel task")
 	}
-	oldwd, err := os.Open(".")
 	if err != nil {
```
Comment (pellared) [inline a.go:313]: **Lifecycle failure can land after the runner returns.** `runAndCleanup` releases the lifecycle lock before `Fatal`/`Fatalf`. A background `Chdir` admitted before completion can fail and block in `Logger.Fatal`; meanwhile cleanup drains, marks the task done, and the runner can return `StatusPassed`. Releasing the logger afterward records the failure too late. `Setenv`, `TempDir`, and parallel-task rejection have the same gap. Track admitted helper calls through their complete error-reporting/Goexit path, and make finalization wait for them. A deterministic regression can block an optional `Logger.Fatal`, return the action after invalid `Chdir` reaches it, and assert the runner waits and ultimately fails.

File: a.go [hunk 5]
```diff
@@ -238,109 +256,196 @@ func (a *A) Helper() {
...
+		a.resourceMu.Lock()
+		defer a.resourceMu.Unlock()
+
+		oldwd, openErr := os.Open(".")
+		if openErr != nil {
+			err = openErr
+			return cleanupEntry{}
+		}
+		restoreDir := func() {
+			if e := oldwd.Chdir(); e != nil {
```
Comment (pellared) [inline a.go]: **Close `oldwd` even when restoration fails.** This branch panics before `oldwd.Close`, leaking the directory descriptor on the exact recovery failure path. Capture the `Chdir` error, close the descriptor, and then panic if restoration failed.

File: a.go [hunk 6]
```diff
@@ -238,109 +256,196 @@ func (a *A) Helper() {
...
+			return cleanupEntry{}
+		}
+		var cleanupErr error
+		return cleanupEntry{
+			resource: true,
+			fn: func() {
+				cleanupErr = os.RemoveAll(dir)
+			},
+			after: func() {
+				if cleanupErr != nil {
```
Comment (pellared) [inline a.go]: **The newly introduced error paths are not covered and both required Codecov checks fail.** Patch coverage is currently 89.31% (14 changed lines missing/partial) and project coverage dropped 0.21%. Please add deterministic coverage for the creation/fatal and cleanup-reporting paths (and the analogous new `Chdir` branches where practical), or simplify the branching enough to restore clean required checks.

File: a_test.go [hunk 7]
```diff
@@ -211,6 +211,167 @@ func TestA_WithContext_concurrent_fail_derived(t *testing.T) {
...
+	}
+}
+
+func TestA_Setenv_concurrent_with_completion(t *testing.T) {
+	const (
+		attempts  = 1000
+		envKey    = "GOYEK_CONCURRENT_COMPLETION_ENV"
+		wantPanic = "Setenv called after task cleanup has completed"
+	)
+	defer func() {
```
Comment (pellared) [inline a_test.go]: **Preserve the caller's environment.** This test unconditionally unsets a process-wide key before and after the loop, so it destroys a pre-existing value. Snapshot `LookupEnv` and restore the original value (or absence) in the deferred cleanup, as the post-completion test already does.

**Review by pellared — COMMENTED** [review]
Second review pass found another lifecycle gap and two cleanup/test issues. The new `WithContext` completion behavior should also be included in the changelog and final PR description. Required Codecov project and patch checks are currently failing, so this round also needs enough deterministic error-path coverage to make all checks green.

File: a_test.go [hunk 8]
```diff
@@ -211,6 +211,266 @@ func TestA_WithContext_concurrent_fail_derived(t *testing.T) {
...
+	case ctx = <-actionReturned:
+	case <-timeout.C:
+		t.Fatal("failed Chdir call did not reach Logger.Fatal")
+	}
+	select {
+	case <-ctx.Done():
+	case <-timeout.C:
+		t.Fatal("task cleanup did not start")
+	}
+	select {
```
Comment (pellared) [inline a_test.go]: **This does not reliably prove the runner stayed blocked.** `ctx.Done` closes at cleanup start, and the immediate `default` can win before a broken runner has simply finished draining/scheduling. Once `release` is closed, that broken implementation can record the failure and pass the test. Keep the logger blocked for a bounded observation interval (or add an internal observation seam) before releasing it. Also make the invalid path a guaranteed-missing child of `t.TempDir()` instead of relying on a checkout-relative literal not existing.

**Review by pellared — COMMENTED** [review]
Third review pass found no remaining implementation or concurrency defect: admission, resource setup/restoration, terminal closure, active-call waiting, cleanup panic recursion, and derived-A sharing are sound. The remaining issues are test determinism/portability and the still-failing required patch-coverage check.

File: a_test.go [hunk 9]
```diff
@@ -393,6 +653,49 @@ func TestA_TempDir(t *testing.T) {
...
+	missingTempDir := filepath.Join(t.TempDir(), "missing")
+	got := goyek.NewRunner(func(a *goyek.A) {
+		a.Setenv("TMPDIR", missingTempDir)
+		a.TempDir()
+	})(goyek.Input{})
+
+	assertEqual(t, got.Status, goyek.StatusFailed, "should fail when the temporary directory cannot be created")
+}
+
+func TestA_TempDir_cleanup_error(t *testing.T) {
```
Comment (pellared) [inline a_test.go]: **These permission-based error tests fail when the test process can bypass mode bits.** Confirmed as UID 0: this test and both new `Chdir` permission tests all fail because `chmod(0)` does not make the operations fail. Probe the required permission behavior and skip when it is unavailable (covering root/capability-enabled containers), or use deterministic injected failure seams.

File: a_test.go [hunk 10]
```diff
@@ -535,6 +838,68 @@ func TestA_Chdir(t *testing.T) {
...
+	originalDir, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+	lockedDir := t.TempDir()
+	targetDir := t.TempDir()
+	if err := os.Chdir(lockedDir); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Chmod(lockedDir, 0); err != nil {
```
Comment (pellared) [inline a_test.go]: **Install the cwd-restoration defer before the next failing operation.** After `os.Chdir(lockedDir)` succeeds, a failure from `os.Chmod` calls `t.Fatal` before the defer exists, leaving the process in a changed working directory and contaminating later tests.

File: a.go [hunk 11]
```diff
@@ -238,109 +260,211 @@ func (a *A) Helper() {
...
-				a.Fatal(err)
-			}
+
+	// Add is serialized with the transition to done. Once runCleanups starts
+	// waiting, no new call can be added to this WaitGroup.
+	a.activeCalls.Add(1)
+	handedOff := false
+	defer func() {
+		a.mu.Unlock()
+		if !handedOff {
```
Comment (pellared) [inline a.go]: **The required Codecov patch check is still red (95.10% vs 96.07%).** This defensive panic handoff is one uncovered/partial changed path. A small white-box test that makes the internal setup callback panic and verifies the active-call count is released would test an important no-deadlock invariant and should close enough of the seven missing/partial lines to restore the required threshold.

File: a_test.go (hunk 8 above)
Comment (pellared) [reply a_test.go]: Addressed in `33d0256` (included in current head `1e70ea8`): the path is now a guaranteed-missing child of `t.TempDir()`, the test keeps `Logger.Fatal` blocked for a 100 ms observation window after cleanup starts, and a deferred release prevents a stranded goroutine on failure.

File: a_test.go (hunk 9 above)
Comment (pellared) [reply a_test.go]: Addressed in `33d0256`: the tests now run an exact capability probe for opening, `fchdir`-ing to, and removing a mode-000 directory. They skip when any required denial is unavailable, including root/capability-enabled environments.

File: a_test.go (hunk 10 above)
Comment (pellared) [reply a_test.go]: Addressed in `33d0256`: cwd restoration is now deferred immediately after the successful `os.Chdir`, before `os.Chmod` can fail.

File: a.go (hunk 11 above)
Comment (pellared) [reply a.go]: Addressed in `33d0256` by simplifying the implementation: setup and cleanup registration already run while holding `a.mu`, so `activeCalls.Add(1)` now happens only after setup succeeds and before unlocking. This removes the rollback branch entirely while preserving Add-before-Wait ordering; local production coverage is 98.0%.
